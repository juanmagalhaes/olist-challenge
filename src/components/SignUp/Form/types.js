// @flow

import { UPDATE, VALIDATE } from './constants';

export type UpdateAction = {
  type: typeof UPDATE,
  payload: { [string]: string }
}

export type ValidateAction = {
  type: typeof VALIDATE,
  payload: {
    password: string,
  },
}

export type Action =
  | UpdateAction
  | ValidateAction;

export type ValidationCriteriaState = {
  msg?: string,
  isValid: boolean,
}

export type PasswordStrengthState = {
  validCount: number,
  minLength: ValidationCriteriaState,
  containsNumber: ValidationCriteriaState,
  containsUppercase: ValidationCriteriaState,
}

export type Validation = {
  passwordStrength: PasswordStrengthState,
}

export type FormState = {
  email: string,
  password: string,
  fullname: string,
  passwordConfirmation: string,
}

export type State = FormState & {
  validation: Validation,
}

