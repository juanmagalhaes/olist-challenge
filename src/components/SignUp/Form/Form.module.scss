$form-padding: 32px 70px 48px;
$form-group-margin-bottom: 24px;

.form {
  text-align: left;
  padding: $form-padding;

  &Group {
    margin-bottom: $form-group-margin-bottom;
  }
}

