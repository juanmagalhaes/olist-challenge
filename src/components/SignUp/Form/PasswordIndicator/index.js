// @flow

export { PasswordIndicatorContainer as PasswordIndicator } from './PasswordIndicator';

