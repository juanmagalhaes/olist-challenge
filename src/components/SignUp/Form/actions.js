// @flow

import { createAction } from 'redux-actions';
import {
  SUBMIT,
  UPDATE,
  VALIDATE,
} from './constants';
import client from '~/api/client';
import type { ValidateAction, UpdateAction } from './types';
import type { ThunkAction } from '~/types/redux';

export const submit = createAction(SUBMIT, async credentials => {
  return await client.submitCredentials(credentials);
});

export function update(field: string, newValue: string): ThunkAction {
  return dispatch => {
    if (field === 'password') {
      const validateAction: ValidateAction = {
        type: VALIDATE,
        payload: { password: newValue },
      };
      dispatch(validateAction);
    }

    const updateAction: UpdateAction = {
      type: UPDATE,
      payload: { [field]: newValue },
    };
    dispatch(updateAction);
  }
}

