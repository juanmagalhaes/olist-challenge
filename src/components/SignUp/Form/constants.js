// @flow

const UPDATE: string = 'UPDATE';
const VALIDATE: string = 'VALIDATE';
const SUBMIT: string = 'SUBMIT';

export {
  UPDATE,
  VALIDATE,
  SUBMIT,
};
