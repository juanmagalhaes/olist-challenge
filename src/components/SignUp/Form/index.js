// @flow

export { FormContainer as Form } from './Form';
export type {
  State,
  Action,
} from './types';

