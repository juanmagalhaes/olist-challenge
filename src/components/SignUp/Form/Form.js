// @flow

import React from 'react';
import { connect } from 'react-redux';
import { Input, Label, Button } from '~/basic-components';
import { PasswordIndicator } from './PasswordIndicator';
import { update, submit } from './actions';
import styles from './Form.module.scss';
import inputStyles from '~/basic-components/Input/Input.module.scss';
import type { FormState } from './types';
import type { ThunkAction } from '~/types/redux';

type Credentials = {
  email: string,
  password: string,
  fullname: string,
}

type Props = FormState & {
  passwordStyle: string,
  update: (string, string) => ThunkAction,
  submit: (Credentials) => ThunkAction,
}

const Form = ({
  update,
  submit,
  email,
  password,
  fullname,
  passwordStyle,
  passwordConfirmation,
}: Props) => {
  const disabled = (!password || !passwordConfirmation)
    || password !== passwordConfirmation;
  return (
    <form
      className={styles.form}
      onSubmit={evt => {
        evt.preventDefault();
        return submit({
          fullname,
          password,
          email,
        })
      }}
    >
      <div className={styles.formGroup}>
        <Label htmlFor="fullname">
          Nome Completo
        </Label>
        <Input
          id="fullname"
          value={fullname}
          autoComplete="on"
          onChange={({ target }) => update('fullname', target.value)}
        />
      </div>
      <div className={styles.formGroup}>
        <Label htmlFor="email">
          E-mail
        </Label>
        <Input
          id="email"
          type="email"
          value={email}
          autoComplete="on"
          onChange={({ target }) => update('email', target.value)}
        />
      </div>
      <div className={styles.formGroup}>
        <Label htmlFor="password">
          Senha
        </Label>
        <Input
          id="password"
          type="password"
          value={password}
          autoComplete="on"
          aria-invalid={passwordStyle !== 'Success'}
          className={inputStyles[`input${passwordStyle}`]}
          onChange={({ target }) => update('password', target.value)}
        />
        <PasswordIndicator />
      </div>
      <div className={styles.formGroup}>
        <Label htmlFor="passwordConfirmation">
          Confirme sua senha
        </Label>
        <Input
          type="password"
          autoComplete="on"
          id="passwordConfirmation"
          value={passwordConfirmation}
          onChange={({ target }) => update('passwordConfirmation', target.value)}
        />
      </div>
      <div>
        <Button
          type="submit"
          color="success"
          disabled={disabled}
        >
          Criar conta
        </Button>
      </div>
    </form>
  )
};

const passwordStyleSelector = ({ validation }) => {
  switch (validation.passwordStrength.validCount) {
    case 1: return 'Danger';
    case 2: return 'Waring';
    case 3: return 'Success';
    default: return 'Default';
  }
}

const mapStateToProps = ({ signUpForm }) => ({
  email: signUpForm.email,
  fullname: signUpForm.fullname,
  password: signUpForm.password,
  passwordStyle: passwordStyleSelector(signUpForm),
  passwordConfirmation: signUpForm.passwordConfirmation,
});

const mapDispatchToProps = (dispatch) => ({
  update: (field, value) => dispatch(update(field, value)),
  submit: async credentials => await dispatch(submit(credentials)),
});

const FormContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(Form);

export { FormContainer, Form };

