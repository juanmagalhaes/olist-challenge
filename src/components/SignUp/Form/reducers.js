// @flow

import { UPDATE, VALIDATE } from './constants';
import type { State, Action } from './types';

export const initialState: State = {
  email: '',
  password: '',
  fullname: '',
  passwordConfirmation: '',
  validation: {
    passwordStrength: {
      minLength: {
        isValid: false,
        msg: 'Pelo menos 6 caracteres',
      },
      containsUppercase: {
        isValid: false,
        msg: 'Pelo menos 1 letra maiúscula',
      },
      containsNumber: {
        isValid: false,
        msg: 'Pelo menos 1 número',
      },
      validCount: 0,
    },
  },
};

function signUpFormReducer(state: State = initialState, action: Action): State {
  const { type, payload } = action;
  switch (type) {
    case UPDATE:
      return {
        ...state,
        ...payload,
      };
    case VALIDATE:
      const { password } = payload,
        minLength = password.length >= 6,
        containsUppercase = /[A-Z]/g.test(password),
        containsNumber = /\d/.test(password),
        validCount = [
          minLength,
          containsUppercase,
          containsNumber,
        ].reduce((acc, criteria) => criteria ? ++acc : acc, 0);

      return {
        ...state,
        validation: {
          passwordStrength: {
            minLength: {
              ...state.validation.passwordStrength.minLength,
              isValid: minLength,
            },
            containsUppercase: {
              ...state.validation.passwordStrength.containsUppercase,
              isValid: containsUppercase,
            },
            containsNumber: {
              ...state.validation.passwordStrength.containsNumber,
              isValid: containsNumber,
            },
            validCount: validCount,
          },
        },
      };
    default:
      return state;
  }
}

export default signUpFormReducer;

