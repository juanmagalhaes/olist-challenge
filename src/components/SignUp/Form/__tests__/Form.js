import React from 'react';
import { Form } from '../Form';

const updateMock = jest.fn();
const submitMock = jest.fn();
const shallowWrapper = props => shallow(<Form {...props} />);
const getWrapper = () => shallowWrapper({
  update: updateMock,
  submit: submitMock,
  email: 'foo@bar.com',
  fullname: 'John Doe',
  password: 'secret123',
  passwordStyle: 'Default',
  passwordConfirmation: 'secret123',
});

test('Form renders', () => expect(getWrapper()).toMatchSnapshot());

test('Form updates', () => {
  const wrapper = getWrapper();
  const inputs = wrapper.find('Input');
  const newValues = [
    'John Doe',
    'foo@bar.com',
    'top secret',
    'confirm top secret'
  ];
  const fields = ['fullname', 'email', 'password', 'passwordConfirmation'];

  inputs.forEach((input, idx) => {
    input.simulate('change', {
      target: { value: newValues[idx] },
    });
    expect(updateMock).lastCalledWith(fields[idx], newValues[idx]);
  });
});

test('Form submits', () => {
  const wrapper = getWrapper();
  const form = wrapper.find('form');
  const preventDefaultMock = jest.fn();
  form.simulate('submit', {
    preventDefault: preventDefaultMock,
  });
  expect(preventDefaultMock).toHaveBeenCalledTimes(1);
  expect(submitMock).toHaveBeenCalledWith({
    fullname: 'John Doe',
    password: 'secret123',
    email: 'foo@bar.com',
  });
});

