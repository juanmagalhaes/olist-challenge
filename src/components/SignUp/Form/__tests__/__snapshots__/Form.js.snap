// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="formGroup"
  >
    <Label
      htmlFor="fullname"
    >
      Nome Completo
    </Label>
    <Input
      autoComplete="on"
      id="fullname"
      onChange={[Function]}
      value="John Doe"
    />
  </div>
  <div
    className="formGroup"
  >
    <Label
      htmlFor="email"
    >
      E-mail
    </Label>
    <Input
      autoComplete="on"
      id="email"
      onChange={[Function]}
      type="email"
      value="foo@bar.com"
    />
  </div>
  <div
    className="formGroup"
  >
    <Label
      htmlFor="password"
    >
      Senha
    </Label>
    <Input
      aria-invalid={true}
      autoComplete="on"
      className="inputDefault"
      id="password"
      onChange={[Function]}
      type="password"
      value="secret123"
    />
    <Connect(PasswordIndicator) />
  </div>
  <div
    className="formGroup"
  >
    <Label
      htmlFor="passwordConfirmation"
    >
      Confirme sua senha
    </Label>
    <Input
      autoComplete="on"
      id="passwordConfirmation"
      onChange={[Function]}
      type="password"
      value="secret123"
    />
  </div>
  <div>
    <Button
      color="success"
      disabled={false}
      type="submit"
    >
      Criar conta
    </Button>
  </div>
</form>
`;
